Bitvector issue with `--type-system-refresh`

### Dafny version

Latest nightly

### Code to produce this issue

```dafny
newtype Uint8 = x: int | 0 <= x < 256

type CustomBv16 = bv16

method Main() {
  var x: Uint8 := 2;
  // Correct rendering of expressions that can interpret the first '<' as a generic argument
  var y: bv16 := (x as CustomBv16) << 2;
}
```

### Command to run and resulting output

```code
> dafny run --target=rs --enforce-determinism --type-system-refresh 

...
error[E0277]: no implementation for `u16 << DafnyInt`
  --> src\conversions.rs:57:41
   |
57 |             let mut y: u16 = (x as u16) << int!(2);
   |                                         ^^ no implementation for `u16 << DafnyInt`
   |
   = help: the trait `Shl<DafnyInt>` is not implemented for `u16`
   = help: the following other types implement trait `Shl<Rhs>`:
             `&u16` implements `Shl<&i128>`
             `&u16` implements `Shl<&i16>`
             `&u16` implements `Shl<&i32>`
             `&u16` implements `Shl<&i64>`
             `&u16` implements `Shl<&i8>`
             `&u16` implements `Shl<&isize>`
             `&u16` implements `Shl<&u128>`
             `&u16` implements `Shl<&u32>`
```

### What happened?

This bug appeared since #5653 
I nailed it down to a single line that needs to be updated. Fix will be provided soon.

### What type of operating system are you experiencing the problem on?

Windows

--- a/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
+++ b/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
@@ -44,7 +44,7 @@ public class BitvectorOptimizationVisitor : BottomUpVisitor {
   }
 
   protected override void VisitOneExpr(Expression expr) {
-    if (expr.Type is BitvectorType bvType) {
+    if (expr.Type.AsBitVectorType is {} bvType) {
 
       if (expr is BinaryExpr binExpr && IsShiftOp(binExpr.Op)) {
         binExpr.E1 = ShrinkBitVectorShiftAmount(binExpr.E1, bvType);

--- a/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
+++ b/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
@@ -44,7 +44,7 @@ public class BitvectorOptimizationVisitor : BottomUpVisitor {
   }
 
   protected override void VisitOneExpr(Expression expr) {
-    if (expr.Type.AsBitVectorType is { } bvType) {
+    if (expr is { Type.AsBitVectorType: { } bvType }) {
 
       if (expr is BinaryExpr binExpr && IsShiftOp(binExpr.Op)) {
         binExpr.E1 = ShrinkBitVectorShiftAmount(binExpr.E1, bvType);

--- a/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
+++ b/Source/DafnyCore/Rewriters/BitvectorOptimization.cs
@@ -44,7 +44,7 @@ public class BitvectorOptimizationVisitor : BottomUpVisitor {
   }
 
   protected override void VisitOneExpr(Expression expr) {
-    if (expr.Type.AsBitVectorType is {} bvType) {
+    if (expr.Type.AsBitVectorType is { } bvType) {
 
       if (expr is BinaryExpr binExpr && IsShiftOp(binExpr.Op)) {
         binExpr.E1 = ShrinkBitVectorShiftAmount(binExpr.E1, bvType);

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/rust/conversions.dfy
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/comp/rust/conversions.dfy
@@ -1,4 +1,6 @@
 // NONUNIFORM: Rust-specific tests
+// RUN: %baredafny run --target=rs --enforce-determinism --type-system-refresh "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
 // RUN: %baredafny run --target=rs --enforce-determinism "%s" > "%t"
 // RUN: %diff "%s.expect" "%t"
 
@@ -27,10 +29,12 @@ method TestCompile(input: Option<Uint8>) returns (output: Option<Uint8>) {
   return None;
 }
 
+type CustomBv16 = bv16
+
 method Main() {
   var x: Uint8 := 2;
   // Correct rendering of expressions that can interpret the first '<' as a generic argument
-  var y: bv16 := (x as bv16) << 2;
+  var y: bv16 := (x as CustomBv16) << 2;
   expect y == 8;
   expect (x as bv16) < 3;
   expect (x as bv16) <= 3;

--- a/docs/dev/news/6196.fix
+++ b/docs/dev/news/6196.fix
@@ -1,0 +1,1 @@
+No more compilation issue when using bitvector shifts in Rust with the new resolver

